### PR TITLE
Add admin monitor to warn users that dangerous permissions are being deprecated

### PR DESCRIPTION
**Pull request by mikecirioli**
The "dangerous" permissions `RUN_SCRIPTS`, `CONFIGURE_UPDATECENTER`, and `UPLOAD_PLUGINS` are being deprecated, so this PR updates the Admin Monitor that is displayed if the user has enabled the use of dangerous permissions via the system property `hudson.security.GlobalMatrixAuthorizationStrategy.dangerousPermissions`.

Additionally, this also fixes the broken layout in the existing admin monitor.

- update to existing admin monitor that is shown if `hudson.security.GlobalMatrixAuthorizationStrategy.dangerousPermissions` is not enabled && their are users with the dangerous permissions
Screen Shot 2020-02-21 at 2 57 16 PM

NOTE: The jenkins version shown is a placeholder and may need to be updated depending on when the deprecation actually lands

- This PR is dependent on https://github.com/jenkinsci/jenkins/pull/4365 being merged and released

### CODE REVIEW

File: src/main/java/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+import hudson.Extension;
+import hudson.model.AdministrativeMonitor;
+import hudson.util.HttpResponses;
+import jenkins.model.Jenkins;
+
+/**
+ * Administrative monitor that shows up when 'dangerous' permissions are enabled.
+ * These permissions are being deprecated and this monitor informs the user of that
```
Comment (aHenryJard) [inline src/main/java/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor.java]:
```suggestion
 * These permissions are being deprecated and this monitor informs the user of that.
```

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.groovy [hunk 2]
```diff
@@ -25,21 +25,23 @@
...
+
     a(href: "https://jenkins.io/redirect/dangerous-permissions") {
         text(_("Learn more"))
     }
-}
 
-form(method: "post", action: "${rootURL}/${it.url}/act") {
-    div {
-        f.submit(name: 'yes', value: _("Review Permissions"))
+    form(method: "post", action: "${rootURL}/${foo.url}/act") {
```
Comment (jtnord) [inline src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.groovy]: ? what's wrong it `it`

File: src/main/resources/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor/message.groovy [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package hudson.security.DeprecatedDangerousMatrixPermissionsAdministrativeMonitor
+
+def f = namespace(lib.FormTagLib)
+def foo = it
```
Comment (jtnord) [inline src/main/resources/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor/message.groovy]: ?

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties [hunk 4]
```diff
@@ -1,8 +1,10 @@
...
+blurb = Your jenkins is currently configured to enable granting at least one of the permissions that can be used to escalate permissions and execute arbitrary code to users without Overall/Administer permission: \
   <ul> \
     <li>Run Scripts</li> \
     <li>Configure Update Center</li> \
     <li>Upload Plugins</li> \
   </ul> \
-  These permissions will no longer actually be granted to users without Overall/Administer permission for security reasons. \
-  It is recommended that you remove these permissions, or grant Overall/Administer to affected users or groups:
\ No newline at end of file
+  <br\> \
```
Comment (jtnord) [inline src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties]: I think you still need to say the above as well as that they are to be removed.  the plugin could be installed on an older version of Jenkins and then the user would not know why `bob` is no longer able to `RUN_SCRIPTS`

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.groovy [hunk 5]
```diff
@@ -25,21 +25,23 @@
 package hudson.security.DangerousMatrixPermissionsAdministrativeMonitor
 
 def f = namespace(lib.FormTagLib)
-
-div(class: "error") {
+def foo = it
```
Comment (aHenryJard) [inline src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.groovy]: Can you use an explicit variable name ? I don't know what foo is just by reading it. Why is it needed, isn't `it` working ?

File: src/main/java/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 CloudBees, Inc.
```
Comment (aHenryJard) [inline src/main/java/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor.java]: I don't know how license works but shouldn't it be 2020 instead ?

File: src/main/resources/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor/message.groovy [hunk 7]
```diff
@@ -0,0 +1,47 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
```
Comment (aHenryJard) [inline src/main/resources/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor/message.groovy]: I don't know how license works but shouldn't it be 2020 instead ?

File: src/main/java/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        if (!(j.getAuthorizationStrategy() instanceof GlobalMatrixAuthorizationStrategy)) {
+            return Collections.emptyList();
+        }
+
+        List<String> sids = new ArrayList<>();
+
+        GlobalMatrixAuthorizationStrategy strategy = (GlobalMatrixAuthorizationStrategy) j.getAuthorizationStrategy();
+
+        List<String> allSidsPlusAnon = new ArrayList<>(strategy.getAllSIDs());
+        allSidsPlusAnon.add("anonymous");
```
Comment (aHenryJard) [inline src/main/java/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor.java]: Do you know if there is a constant for this specific user ?

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.groovy (hunk 5 above)
Comment (mikecirioli) [reply src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.groovy]: Fixed, i had to use the variable to deal with a groovy scope issue

File: src/main/java/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor.java (hunk 8 above)
Comment (mikecirioli) [reply src/main/java/hudson/security/DeprecatedDangerousMatrixPermissionsAdministrativeMonitor.java]: There is (`Acl.ANONYMOUS_USERNAME`) but its annotated as `@NoExternalUse`

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.groovy (hunk 2 above)
Comment (mikecirioli) [reply src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.groovy]: scope was not available (for some reason i don't clearly understand)

File: src/main/java/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor.java [hunk 9]
```diff
@@ -48,7 +48,8 @@
 public class DangerousMatrixPermissionsAdministrativeMonitor extends AdministrativeMonitor {
     @Override
     public boolean isActivated() {
-        return !GlobalMatrixAuthorizationStrategy.ENABLE_DANGEROUS_PERMISSIONS && !getSidsWithDangerousPermissions().isEmpty();
+        return ((!GlobalMatrixAuthorizationStrategy.ENABLE_DANGEROUS_PERMISSIONS && !getSidsWithDangerousPermissions().isEmpty()) ||
```
Comment (jtnord) [inline src/main/java/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor.java]: `getSidsWithDangerousPermissions().isEmpty()` seems quite expensive to check on every request?

**Comment by daniel-beck** [comment]
What is the relation of this PR to https://github.com/jenkinsci/matrix-auth-plugin/pull/73?

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties [hunk 10]
```diff
@@ -1,8 +1,13 @@
-blurb = Your authorization strategy is currently configured to grant at least one of the permissions that can be used to escalate permissions and execute arbitrary code to users without Overall/Administer permission: \
+blurb = Your jenkins is currently configured to enable granting at least one of the permissions that can be used to escalate permissions and execute arbitrary code to users without Overall/Administer permission: \
```
Comment (daniel-beck) [inline src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties]:
```suggestion
blurb = Your Jenkins is currently configured to enable granting at least one of the permissions that can be used to escalate permissions and execute arbitrary code to users without Overall/Administer permission: \
```

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties [hunk 11]
```diff
@@ -1,8 +1,13 @@
...
   <ul> \
     <li>Run Scripts</li> \
     <li>Configure Update Center</li> \
     <li>Upload Plugins</li> \
   </ul> \
+  <br\> \
```
Comment (daniel-beck) [inline src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties]:
```suggestion
  <br/> \
```

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties [hunk 12]
```diff
@@ -1,8 +1,13 @@
...
+  <b>These permissions have been marked as DEPRECATED as of Jenkins v2.222 and will be removed completely in a future release.</b><br\> \
+   \
+  <br\> \
```
Comment (daniel-beck) [inline src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties]:
```suggestion
  <br/> \
```

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties [hunk 13]
```diff
@@ -1,8 +1,13 @@
...
   These permissions will no longer actually be granted to users without Overall/Administer permission for security reasons. \
-  It is recommended that you remove these permissions, or grant Overall/Administer to affected users or groups:
\ No newline at end of file
+  It is recommended that you remove these permissions, or grant Overall/Administer to affected users or groups: \
+  <br>
```
Comment (daniel-beck) [inline src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties]:
```suggestion
```

File: src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties [hunk 14]
```diff
@@ -1,8 +1,13 @@
-blurb = Your authorization strategy is currently configured to grant at least one of the permissions that can be used to escalate permissions and execute arbitrary code to users without Overall/Administer permission: \
+blurb = Your jenkins is currently configured to enable granting at least one of the permissions that can be used to escalate permissions and execute arbitrary code to users without Overall/Administer permission: \
   <ul> \
     <li>Run Scripts</li> \
     <li>Configure Update Center</li> \
     <li>Upload Plugins</li> \
   </ul> \
+  <br\> \
+  <b>These permissions have been marked as DEPRECATED as of Jenkins v2.222 and will be removed completely in a future release.</b><br\> \
```
Comment (daniel-beck) [inline src/main/resources/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor/message.properties]:
```suggestion
  <b>These permissions have been deprecated in Jenkins 2.222 and will be removed completely in a future release.</b><br/> \
```

File: src/main/java/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor.java (hunk 9 above)
Comment (mikecirioli) [reply src/main/java/hudson/security/DangerousMatrixPermissionsAdministrativeMonitor.java]: That is how the monitor was originally implemented....

`return !GlobalMatrixAuthorizationStrategy.ENABLE_DANGEROUS_PERMISSIONS && !getSidsWithDangerousPermissions().isEmpty(); `

I assume it was originally implemented this way to inform users who had previously used these permissions before they became disabled by default.  Does it make more sense to just show it to users who are still using these permissions now?

**Comment by mikecirioli** [comment]
> What is the relation of this PR to #73?

@daniel-beck - #73 is now closed, this PR replaces it

**Comment by timja** [comment]
Superseded by https://github.com/jenkinsci/matrix-auth-plugin/pull/80 by the looks of it

**Comment by daniel-beck** [comment]
I went with #80 instead.

**Pull request closed without merging**